x-pack/plugins/security/server/authentication/providers/saml.ts

```diff
@@ -113,7 +124,15 @@ export class SAMLAuthenticationProvider extends BaseAuthenticationProvider {
   ) {
     this.logger.debug('Trying to perform a login.');
 
-    if (attempt.step === SAMLLoginStep.RedirectURLFragmentCaptured) {
+    // It may happen that Kibana is re-configured to use different realm for the same provider name,
+    // we should clear such session an log user out.
+    if (state?.realm && state.realm !== this.realm) {
```

Comment:
__note:__ this is check is the main/only reason why I started to store `realm` in the session.